parser/parser.go

```diff
@@ -6,127 +6,168 @@ import (
+	br := bufio.NewReader(r)
+	for {
+		r, _, err := br.ReadRune()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
 
-	scanner := bufio.NewScanner(reader)
-	scanner.Buffer(make([]byte, 0, bufio.MaxScanTokenSize), bufio.MaxScanTokenSize)
-	scanner.Split(scanModelfile)
-	for scanner.Scan() {
-		line := scanner.Bytes()
+		if _, err := cmd.WriteRune(r); err != nil {
+			return nil, err
+		}
 
-		fields := bytes.SplitN(line, []byte(" "), 2)
-		if len(fields) == 0 || len(fields[0]) == 0 {
+		// trim leading whitespace
+		if (space(r) || newline(r)) && b.Len() == 0 {
 			continue
 		}
 
-		switch string(bytes.ToUpper(fields[0])) {
-		case "FROM":
-			command.Name = "model"
-			command.Args = string(bytes.TrimSpace(fields[1]))
-			// copy command for validation
-			modelCommand = command
-		case "ADAPTER":
-			command.Name = string(bytes.ToLower(fields[0]))
-			command.Args = string(bytes.TrimSpace(fields[1]))
-		case "LICENSE", "TEMPLATE", "SYSTEM", "PROMPT":
-			command.Name = string(bytes.ToLower(fields[0]))
-			command.Args = string(fields[1])
-		case "PARAMETER":
-			fields = bytes.SplitN(fields[1], []byte(" "), 2)
-			if len(fields) < 2 {
-				return nil, fmt.Errorf("missing value for %s", fields)
+		switch s {
+		case stateName, stateParameter:
+			if alpha(r) || number(r) || (r == '_' && s == stateParameter) {
+				if _, err := b.WriteRune(r); err != nil {
+					return nil, err
+				}
+			} else if space(r) {
+				cmd.Name = strings.ToLower(b.String())
+				b.Reset()
+
+				if cmd.Name == "from" {
+					cmd.Name = "model"
+				}
+
+				switch cmd.Name {
+				case "parameter":
+					s = stateParameter
+				case "message":
+					s = stateMessage
+				default:
+					s = stateArgs
+				}
+			} else if newline(r) {
+				return nil, fmt.Errorf("missing value for [%s]", b.String())
+			} else {
+				return nil, fmt.Errorf("unexpected rune %q for state %d", r, s)
 			}
-
-			command.Name = string(fields[0])
-			command.Args = string(bytes.TrimSpace(fields[1]))
-		case "EMBED":
-			return nil, fmt.Errorf("deprecated command: EMBED is no longer supported, use the /embed API endpoint instead")
-		case "MESSAGE":
-			command.Name = string(bytes.ToLower(fields[0]))
-			fields = bytes.SplitN(fields[1], []byte(" "), 2)
-			if len(fields) < 2 {
-				return nil, fmt.Errorf("should be in the format <role> <message>")
+		case stateArgs:
+			if r == '"' && b.Len() == 0 {
+				quotes++
+				s = stateMultiline
+			} else if newline(r) {
+				cmd.Args += b.String()
+				b.Reset()
+
+				cmds = append(cmds, cmd)
+				cmd = Command{}
+				s = stateName
+			} else {
+				if _, err := b.WriteRune(r); err != nil {
+					return nil, err
+				}
 			}
-			if !slices.Contains([]string{"system", "user", "assistant"}, string(bytes.ToLower(fields[0]))) {
-				return nil, fmt.Errorf("role must be one of \"system\", \"user\", or \"assistant\"")
+		case stateMultiline:
+			if r == '"' && b.Len() == 0 {
+				quotes++
+			} else if r == '"' {
```

Comment:
This will get tripped up on something like:

MESSAGE assistant """You are a "thing""""

I think you need to look for something like \" to allow the user to unescape double quotes after the first char. I can see this being a problem w/ the `/save` functionality in the REPL.

Reply:
multiline is currently broken